### ISSUE TITLE
violates the constraint of type parameter 'TUser'

### ISSUE THREAD

**Issue by seven1986** [aspnet/Identity#1319]
GenericArguments[0], 'WebApplication1.Models.AppUser', on 'Microsoft.AspNetCore.Identity.EntityFrameworkCore.UserStore`9[TUser,TRole,TContext,TKey,TUserClaim,TUserRole,TUserLogin,TUserToken,TRoleClaim]' violates the constraint of type parameter 'TUser'.

I spend much time, still can not sovle this issue, please help, thanks a lot!

![image](https://user-images.githubusercontent.com/20808494/28262761-db82dce6-6b16-11e7-9320-e5dcad7f95ea.png)

**AppTables.cs**
```csharp
    public class AppUser : IdentityUser<long,AppUserClaim,AppUserRole,AppUserLogin>   { }

    public class AppUserClaim : IdentityUserClaim<long>  { }

    public class AppUserToken : IdentityUserToken<long> { }

    public class AppUserLogin: IdentityUserLogin<long> { }

    public class AppUserRole : IdentityUserRole<long> { }

    public class AppRole : IdentityRole<long,AppUserRole,AppRoleClaim>  {}

    public class AppRoleClaim : IdentityRoleClaim<long> { }
```

**AppDbContext.cs**
```csharp
 public class AppDbContext : IdentityDbContext<
        AppUser,
        AppRole,
        long,
        AppUserClaim,
        AppUserRole,
        AppUserLogin,
        AppRoleClaim,
        AppUserToken>
    {
        public AppDbContext(DbContextOptions<AppDbContext> options)
            : base(options)
        {

        }

        protected override void OnModelCreating(ModelBuilder builder)
        {
            base.OnModelCreating(builder);

            builder.Entity<AppUser>().ToTable("User");
            builder.Entity<AppRole>().ToTable("Role");
            builder.Entity<AppUserClaim>().ToTable("UserClaim");
            builder.Entity<AppUserLogin>().ToTable("UserLogin");
            builder.Entity<AppUserRole>().ToTable("UserRole");
            builder.Entity<AppUserToken>().ToTable("UserToken");
            builder.Entity<AppRoleClaim>().ToTable("RoleClaim");
        }
    }
```

**Startup.cs**
```csharp
services.AddIdentity<AppUser, AppRole>()
            .AddEntityFrameworkStores<AppDbContext,long>()
            .AddDefaultTokenProviders();
```

**Comment by series0ne** [comment]
I believe I've had the same problem, documented [here ](https://stackoverflow.com/questions/45127666/aspnetcore-identity-custom-implementation-not-working)on StackOverflow.

The best answer given was

The constraint is `where TRole : IdentityRole<TKey, TUserRole, IdentityRoleClaim<TKey>>`. Since this is class constraint, there is no co/contravariance. IMO they simply forgot to add `TRoleClaim` generic argument.

the ASP.NET Core Identity team should investigate this...seems like this should be a simple fix.

**Comment by series0ne** [comment]
In fact, the issue I posted above isn't exactly the same, but still closely related to this issue. The problem above is that custom identity types for `TUser`, `TRole` etc. can't be used with

`.AddEntityFrameworkStores<AppDbContext,long>()` and `.AddDefaultTokenProviders();` You need to implement all of this yourself.

See this article
https://github.com/aspnet/Identity/issues/1082

**Comment by HaoK** [comment]
This was fixed in 1.1, dupe of https://github.com/aspnet/Identity/issues/855

**Comment by KiselN** [comment]
@HaoK TRole issue doesn't seem to be fixed in 1.1.1
https://github.com/aspnet/Identity/blob/rel/1.1.1/src/Microsoft.AspNetCore.Identity.EntityFrameworkCore/UserStore.cs#L164
